pytorch_lightning/overrides/distributed.py

```diff
@@ -104,6 +106,103 @@ def __iter__(self) -> Iterator[List[int]]:
+                " is True and you are using distributed training. Either provide `__len__`"
+                " method in your sampler, remove it from DataLoader or set `replace_sampler_ddp=False`"
+                " if you want to handle distributed sampling yourself."
+            )
+        if len(sampler) == float("inf"):
+            raise MisconfigurationException(
+                "You seem to have configured a sampler in your DataLoader which"
+                " does not provide finite `__len__` method. The sampler was about to be"
+                " replaced by `DistributedSamplerWrapper` since `replace_sampler_ddp`"
+                " is True and you are using distributed training. Either provide `__len__`"
+                " method in your sampler which returns a finite number, remove it from DataLoader"
+                " or set `replace_sampler_ddp=False` if you want to handle distributed sampling yourself."
+            )
+            raise TypeError("The given sampler must be bounded.")
+        self._sampler = sampler
+        self._sampler_list: Optional[List[Any]] = None
+
+    def __getitem__(self, index: int) -> Any:
+        """Gets element of the dataset.
+
+        Args:
+            index: index of the element in the dataset
+        Returns:
+            Single element by index
+        """
+        if self._sampler_list is None:
+            self._sampler_list = list(self._sampler)
+        return self._sampler_list[index]
+
+    def __len__(self) -> int:
+        """
+        Returns:
+            int: length of the dataset
+        """
+        return len(self._sampler)
+
+
+def _subsample_dataset(dataset: Dataset, iterator_: Iterator) -> Iterator:
+    indexes_of_indexes = list(iterator_)
+    num_indices = len(indexes_of_indexes)
+    indices: Tuple = ()
+    if num_indices:
+        indices = itemgetter(*indexes_of_indexes)(dataset)
+        if num_indices == 1:
+            indices = (indices,)  # avoid splitting
+    return iter(indices)
+
+
+class DistributedSamplerWrapper(DistributedSampler):
+    """Wrapper over `Sampler` for distributed training. Allows you to use any sampler in distributed mode.
+
+    It is especially useful in conjunction with :class:`torch.nn.parallel.DistributedDataParallel`.
+    In such case, each process can pass a ``DistributedSamplerWrapper`` instance as a ``DataLoader``
+    sampler, and load a subset of subsampled data of the original dataset that is exclusive to it.
+
+    .. note::
+        Sampler is assumed to be of constant size.
+
+    Adapted from https://github.com/catalyst-team/catalyst/blob/v22.04/catalyst/data/sampler.py#L499
+    """
+
+    def __init__(self, sampler: Union[Sampler, Iterable], *args: Any, **kwargs: Any) -> None:
+        super().__init__(_DatasetFromSampler(sampler), *args, **kwargs)
+        self.sampler = sampler
+
+    def __iter__(self) -> Iterator:
+        self.dataset = _DatasetFromSampler(self.sampler)
```

Comment:
```suggestion
        super().__init__(_DatasetFromSampler(sampler), *args, **kwargs)
        self.sampler = sampler

    def __iter__(self) -> Iterator:
        self.dataset = _DatasetFromSampler(self.sampler)
```
just wondering, will they be two different instances: `_DatasetFromSampler(self.sampler)`
why is `self.dataset` required to be instantiated at each iter?

Reply:
Yes, they will be two different instances. We don't really care about the instance in the `__init__`, since in `__init__` we only care about `__len__`.

But we want to create a fresh iterator everytime we restart the iterator, since the values can change depending on things like seed, etc